paddle/fluid/platform/init.cc

```diff
@@ -132,14 +138,36 @@ void InitDevices(bool init_p2p, const std::vector<int> devices) {
       continue;
     }
 
-    places.emplace_back(platform::CUDAPlace(devices[i]));
+    places.insert(platform::CUDAPlace(devices[i]));
   }
   if (init_p2p) {
     InitP2P(devices);
   }
-  places.emplace_back(platform::CPUPlace());
-  platform::DeviceContextPool::Init(places);
+  places.insert(platform::CPUPlace());
+  std::vector<platform::Place> devs(places.begin(), places.end());
+  platform::DeviceContextPool::Init(devs);
   platform::DeviceTemporaryAllocator::Init();
+
+#if defined(PADDLE_WITH_CUDA) && !defined(_WIN32) && \
+    !defined(PADDLE_ON_INFERENCE)
+  std::vector<platform::Place> cuda_devs;
+  for (auto &p : places) {
+    if (platform::is_gpu_place(p)) {
+      cuda_devs.push_back(p);
+    }
+  }
+  std::unique_ptr<platform::NCCLContextMap> nccl_ctxs;
+  nccl_ctxs.reset(new platform::NCCLContextMap(cuda_devs));
```

Comment:
Maybe better to initialize at `ParallelExecutor` to be compatible with distributed mode, which needs to pass global number of devices in the cluster. In `init.cc` it can only get local device count.

Reply:
Done. Thanks!